pkg/transport/generic_transport_controller.go

```diff
@@ -19,16 +19,21 @@ package transport
 import (
 	"context"
 	"fmt"
+	"sync"
 	"time"
 
 	"github.com/go-logr/logr"
+	clusterinformers "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1"
+	clusterlisters "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+	clusterapi "open-cluster-management.io/api/cluster/v1"
```

Comment:
I really think using the clusters objects for storing customization parameters is a wrong decision.

clusters objects may have a lot of events (e.g, for updating periodic liveness status) that most of them are not related to customization and in general it's not very clear why would one use labels/annotations and the precedence order you defined.
labels are typically used for group selection not for configuration values. plus, any change in any cluster labels will trigger re-evaluation of all binding policies (through the binding controller). 

I think that using MC to store the parameters for customization is kind of a mix between non-related points that looks very awkward to me. 

IMO a much more native decision is to implement a very straight forward thin controller that runs in ITS, that watches clusters objects and upon creation of MC creates a matching ConfigMap in its dedicated ns with the same name (we can even define the MC as OwnerReference of the ConfigMap). this controller can ignore update events of MC (cause that's not interesting) and also ignore delete events (ownerReference can delete the configmap).
we can also define that the configmap includes default fields that KS provides like key `name` and value `cluster_name`. (e.g., what Andy wanted for Loki).

customization key value pairs are configuration values and as such, It feels much more native to store it in ConfigMap + it keeps the clear separation between Inventory and Customization parameters.

Reply:
@ezrasilvera @pdettori would be helpful to hear also your thoughts on this point.

Reply:
I went with the simplest thing I could think of.

Reply:
I think using ConfigMap would be much simpler :).
both in terms of code readability, maintenance and clear separation of concerns and in terms of UX which I think would much easier for users. users of k8s know that configmaps are used for storing key-value configuration pairs.

Reply:
do you have a reference for this point from sig-multicluster?
the ideal would be to get aligned with sig-multicluster.
 
intuitively, I think sig-multicluster wouldn't go into this direction of using both labels and annotations with a precedence order. I would be very surprised if that happens.

can we agree on checking how sig-multicluster defining cluster properties and get aligned with that?
It would also help us later in the inventory work for easier alignment with sig-multicluster.

Reply:
I did not mean to suggest that sig-multicluster was going to propagate cluster self-description properties into inventory object labels and annotations; I think it would be into some modeled part of the inventory object. I got this in a verbal discussion in a meeting. There might be a recording. There might be something written down.

Reply:
it was understood from your words that SIG-MC are doing that or going to do that.
if that was the case I'd say we can treat SIG-MC as a standard and align with that.

but after reading the last message I understand that is not the case.
I was trying to google quickly (didn't invest more than 5 minutes) and haven't found any docs about how SIG-MC are going to define properties of clusters. so I don't think at this point there is a standard for this thing and therefore we should do what's right for KS.

btw, on the other hand, the standard in kubernetes in general (not specifically in SIG-MC) for defining key-value configuration pairs is using configmaps. nothing prevents us from using labels + annotations but the standard is definitely configmaps.

I also don't agree with you previous comment:
> I was trying to follow Occam's Razor ("Entities must not be multiplied beyond necessity").
> 
> The ManagedCluster describes the WEC. I think that it is entirely reasonable to look for WEC properties in the labels and annotations of the ManagedCluster. Labels and annotations are open for any use, as far as the Kube metamodel is concerned. Also in this vein: my understanding of the direction that sig-multicluster is going in will see the self-describing cluster properties inside the WEC propagated to the core to appear in the inventory object.

Eventually we need to define key-value properties pairs somewhere. 
one option that you suggested is to use labels/annotations.
another option that I suggest is to use configmaps.
the properties are not multiplied. the fact that it's in a different object doesn't mean we multiply beyond necessity. there is a necessity to separate the things cause using inventory for defining customization parameters feels like a spaghetti code. 
separation of concerns is widely used in software engineering and it doesn't mean we multiply beyond necessity.

Reply:
@MikeSpreitzer @nirrozenbaum My 2 cents:  My original vision was to have a "chain of config inputs" , similar to what is usually done with application configuration. When a key is found in a level the search is aborted and that value is used. 
This allows the possibility to override values, specify values for different criteria, etc..

We can think of multiple such levels also in our case:
1. CM matching a label
2. CM per BP
3. CM per MC
4. etc..
At this time I agree we can just have parameters per MC - but I agree I would personally use a different object and not the MC object itself, for several reasons:
1. Allows flexibility/override between WDSs - E.g., special patter for cluster identifier 
2. Security - we may not allow in the future setting fields on the MC object itself as it is managed by the inventory and may risk the whole environment
3. Clutter: What happen if we want to use many different keys?  (cluster-name, cluster-ip, etc..)

Reply:
@ezrasilvera: A layered or chained approach explicitly contemplates multiple layers and can start with having only one simple layer.

On further thought, I am less enthused about the user experience of using inventory object labels and annotations as the source of parameters. Now that we are using Go language templates, the parameter names have to be valid as identifiers in the Go language. Label and annotation names (keys) commonly have characters (`-`, `.`, `/`) that are not allowed in a Go identifier. We _could_ try to define mapping from labels + annotations to a data structure using only valid Go identifiers. E.g., `foo.com/bar=baz` maps to `map[string]map[string]map[string]string{"foo": map[string]map[string]string{"com": map[string]string{"bar": "baz"}}}`. That would be sort of doable. What would `foo=xyz,foo.bar=abc` map to?

Another possible approach would be to define a mapping of the three problematic characters to ones that are allowed in a Go identifier. Something like mapping `foo-bar.com/baz` to `foo🄼bar🄳com🅂bar`.

Both of these approaches mean that any change to any label or annotation of an inventory object would require re-creation of every ManifestWork that is sensitive to that inventory object's labels and annotations.